File: 637. Average of Levels in Binary Tree/Java/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        Queue<TreeNode> queue = new ArrayDeque<>();
        List<Double> result = new ArrayList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            long sumCurrentLevel = 0;
            int countCurrentLevel = queue.size();
            for (int i = 0; i < countCurrentLevel; i++) {
                TreeNode current = queue.poll();
                sumCurrentLevel += current.val;
                if (current.left != null) {
                    queue.add(current.left);
                }
                if (current.right != null) {
                    queue.add(current.right);
                }
            }
            if (countCurrentLevel != 0) {
                result.add((double) sumCurrentLevel / countCurrentLevel);
            }
        }
        return result;
    }
}

File: 637. Average of Levels in Binary Tree/C++/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode *root) {
        if (root == nullptr) {
            return vector<double>();
        }
        queue<TreeNode *> que;
        vector<double> result;
        que.push(root);
        while (!que.empty()) {
            long sumCurrentLevel = 0;
            int countCurrentLevel = que.size();
            for (int i = 0; i < countCurrentLevel; i++) {
                TreeNode *current = que.front();
                que.pop();
                sumCurrentLevel += current->val;
                if (current->left != nullptr) {
                    que.push(current->left);
                }
                if (current->right != nullptr) {
                    que.push(current->right);
                }
            }
            if (countCurrentLevel != 0) {
                result.push_back((double) sumCurrentLevel / countCurrentLevel);
            }
        }
        return result;
    }
};